oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/endpoint/DefaultOAuth2TokenRequestHeadersConverter.java

```diff
@@ -29,50 +28,54 @@
 import org.springframework.security.oauth2.core.ClientAuthenticationMethod;
 
 /**
- * Utility methods used by the {@link Converter}'s that convert from an implementation of
- * an {@link AbstractOAuth2AuthorizationGrantRequest} to a {@link RequestEntity}
- * representation of an OAuth 2.0 Access Token Request for the specific Authorization
- * Grant.
+ * Default Converter used by the
+ * {@link OAuth2AuthorizationCodeGrantRequestEntityConverter} that convert from an
+ * implementation of an {@link AbstractOAuth2AuthorizationGrantRequest} to a
+ * {@link RequestEntity} representation of an OAuth 2.0 Access Token Request for the
+ * specific Authorization Grant.
  *
+ * @author Peter Eastham
  * @author Joe Grandja
- * @since 5.1
- * @see OAuth2AuthorizationCodeGrantRequestEntityConverter
+ * @since 6.3
  * @see OAuth2ClientCredentialsGrantRequestEntityConverter
  */
-final class OAuth2AuthorizationGrantRequestEntityUtils {
+public class DefaultOAuth2TokenRequestHeadersConverter<T extends AbstractOAuth2AuthorizationGrantRequest>
+		implements Converter<T, HttpHeaders> {
 
-	private static HttpHeaders DEFAULT_TOKEN_REQUEST_HEADERS = getDefaultTokenRequestHeaders();
+	private static final HttpHeaders DEFAULT_TOKEN_HEADERS = getDefaultTokenRequestHeaders();
 
-	private OAuth2AuthorizationGrantRequestEntityUtils() {
+	private boolean encodeClientCredentials = true;
+
+	private static HttpHeaders getDefaultTokenRequestHeaders() {
+		HttpHeaders headers = new HttpHeaders();
+		headers.setAccept(Collections.singletonList(MediaType.APPLICATION_JSON_UTF8));
+		final MediaType contentType = MediaType.valueOf(MediaType.APPLICATION_FORM_URLENCODED_VALUE + ";charset=UTF-8");
```

Comment:
In order to bring this into alignment with the corresponding logic on the reactive side (see `AbstractWebClientReactiveOAuth2AccessTokenResponseClient#populateTokenRequestHeaders`), I wonder if there's anything we can do about this `charset=UTF-8`? For backwards compatibility, it might be nice to make this additional param on the `Content-Type` configurable. If we do this, we can re-use this exact implementation for the reactive side as well. (see opening comment)

Reply:
Removing the `charset=UTF-8` breaks about 19 tests. Do you want me to remove that check from the tests? Or maintain a default of `charset=UTF-8`? (More information on a different conversation)

Reply:
@Crain-32 We don't want to change the behavior for the servlet side (tests should not be changed). But `charset=UTF-8` should not be the default for the new class either. Instead, the `AbstractOAuth2AuthorizationGrantRequestEntityConverter` should be able to easily create an instance of this new class with the addition of `charset=UTF-8` enabled. Then the reactive version would just instantiate and use the class with defaults.

The best way to do this IMHO would be to create a package-private static factory method in this new class that instantiates the class and enables the customized `charset`, which is used by `AbstractOAuth2AuthorizationGrantRequestEntityConverter`. Make sense?

Reply:
@sjohnr it does make sense to use a factory. Would the preferred naming be something like `....HeaderConverterCustomizer`, or `.....HeaderConverterFactory`? The first feels more in line with the general idea of "Customizer" the JWT Client customizer does, where the second one feels more generic to the Java ecosystem.


Follow up question on that though, would it be better to expose the factory/builder within the public scope instead of the class?
Current plan for customization was just having the consumer use `new Default.....()`, but if we're having the internals use a factory, might as well expose the factory instead of the class.

Reply:
> Would the preferred naming be something like `....HeaderConverterCustomizer`, or `.....HeaderConverterFactory`?

It should be a package-private static method, not a class. The name doesn't exactly matter because it will be hidden at the package level. However, `withCharsetUtf8()` would be my preferred name, as it tells the story fairly well.

> Current plan for customization was just having the consumer use `new Default.....()`, but if we're having the internals use a factory, might as well expose the factory instead of the class.

I don't think that we should expose this. The charset customization is only being used to make this change non-breaking. The `charset=UTF-8` variants in `MediaType` are deprecated and should not be used. We don't want to encourage their use, so keeping this customization internal and only used by default implementations provided by the framework is preferred. Said another way, this customization should stay hidden in the framework and only used to prevent breaking tests (e.g. backwards compatibility).